gleam: Add erlang_27 as runtime dependency

bhankas opened the pull request:
While gleam can be run entirely off JS backend, a lot of its intended uses are targeted at back-end where its target is BEAM VM of Erlang.

Current Gleam version requires Erlang/OTP 26 (erlang_26), while nixpkgs still defaults to Erlang/OTP 25 (erlang). This change fixes that.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
gleam: Add erlang_27 as runtime dependency

While gleam can be run entirely off JS backend, a lot of its intended
uses are in the backend where its target is BEAM VM of Erlang.

Current Gleam version requires Erlang/OTP 26 (erlang_26), while
nixpkgs still defaults to Erlang/OTP 25 (erlang). This change fixes
that.

adamcstephens reviewed (commented):
Please undo the formatting, this shouldn't be included in standard PRs. 

The package should take `erlang` as an attribute and if a different version than default is needed it should be set in all-packages. See https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults

bhankas commented:
> Please undo the formatting, this shouldn't be included in standard PRs.
> 
> The package should take `erlang` as an attribute and if a different version than default is needed it should be set in all-packages. See https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults

Thanks for the review!

I reverted formatting changes. I added an input argument, but instead of calling with `erlang_26` from all-packages.nix (which should be done away with, really) I added default value in default.nix itself. IMO this is more discoverable for future updates, but advice welcome.

adamcstephens commented on pkgs/development/compilers/gleam/default.nix:
```diff
@@ -5,9 +5,11 @@
 , git
 , pkg-config
 , openssl
+, erlangVersion ? null
 , Security
 , nix-update-script
 , SystemConfiguration
+, pkgs
```
```suggestion
, erlang
, Security
, nix-update-script
, SystemConfiguration
```

adamcstephens reviewed (changes requested):
all-packages.nix is not going away, and the link I shared explains why the version should be set there and not in this file. What you're proposing here is very non-standard.

bhankas commented:
> all-packages.nix is not going away, and the link I shared explains why the version should be set there and not in this file. What you're proposing here is very non-standard.

Fair points, I've updated the code with suggested changes. Let me know if this works.

adamcstephens reviewed (approved):
Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2067
